### DISCUSSION TITLE
Why `const` is slower than `var<private>` for arrays

### DISCUSSION THREAD

**Discussion by munrocket** [gpuweb/gpuweb#4281] [Q&A] [answered]
This is checked across different devices/OSs and probably not implementation issue.

37 line here explains issue:
https://compute.toys/view/359
Try to change B/W top level declaration / usage in code and check FPS.

Why this happening? Looks like there are no reason to use const at all and override/var<private> is always better.

**Comment by ben-clayton** [comment, marked as answer]
You've found a performance cliff with Tint's implementation of `const`. There's a few unique specifics with your example that triggers this:

* You have a `const` array.
* The `const` array is being indexed with a runtime expression.
* You're using the dynamic index in a deeply nested loop.

In your example, Tint is inlining the `const` array within the scope of the index operation. The reasons for this are:

* A single `const` variable can be used to emit values of different backend types. Instead of `array<i32, 12>(-241, ..., 202)`, you could write `array(-241, ..., 202)` which is an array of `abstract-int` and the single `const` array can be used as an array of `i32`, `u32`, `f32` or `f16`.
* A `const` array of abstract-numerics may hold numerical values that are not representable in MSL, so we don't emit `const` variables until the final outer expression. For example you could write `const C = array(0x1000000000); var<private> v = C[0] >> 32;`
* Because you're indexing the `const` array with a runtime expression, we need to emit the full array and perform the index operation as an instruction. If this were a constant index, then we'd just emit the indexed value.
* Because a `const` array takes the type of its usage, we currently inline the array at the place it is used - in your example, this is within the most nested scope.

We've been assuming that backend compilers would be smart enough to avoid any instructions for generating read-only "local" arrays, but your example has highlighted that many drivers have substantial runtime overhead for these variable arrays.

Tint will probably need to hoist these `const` arrays to a global scope, and deal with the one-`const` to many-usages issue. I'll file a bug to do this.

Thank you for reporting this issue.

**Reply by munrocket** [reply]
Linux backend with AMD also affected to this.

**Comment by dneto0** [comment]
FYI.  On my laptop, changing the 'var's on lines 38 and 39 to 'let' keeps the high performance behaviour.
That's expected, given Ben's explanation about 'const' being the issue specifically.

**Reply by munrocket** [reply]
To get performance drop, L38, L39 should be removed and ws -> W, bs -> B renamed.
After that `var<private>` instead of `const` improving performance as it was.

**Comment by dneto0** [comment]
> Why this happening? Looks like there are no reason to use const at all and override/var is always better.

To answer this second part more explicitly (and assembling bits Ben already mentioned):

- Const array will always be better if its index is const.  Tint will fold that down into just the element itself.  That's not required by WGSL, and other implementations may behave differently.
- 'const' values can use 64bits, so you can do more intermediate calculations in higher fp precision or integer magnitudes, as long as they eventually produce a const value that can fit in the target 32bit (concrete) type.

**Comment by ben-clayton** [comment]
I've created this tint issue to track this: [crbug.com/tint/2032](https://crbug.com/tint/2032)
